CRM/Contribute/Form/AbstractEditPayment.php

```diff
@@ -432,12 +432,14 @@ public function assignProcessors() {
       }
       $this->_processors = array();
       foreach ($this->_paymentProcessors as $id => $processor) {
-        $this->_processors[$id] = ts($processor['name']);
-        if (!empty($processor['description'])) {
-          $this->_processors[$id] .= ' : ' . ts($processor['description']);
-        }
-        if ($processor['is_recur']) {
-          $this->_recurPaymentProcessors[$id] = $this->_processors[$id];
+        if ($processor['is_test'] == ($this->_mode == 'test')) {
```

Comment:
I think the problem is in this function

https://github.com/civicrm/civicrm-core/blob/a160bb08153122ed40338b62aa68cc5abc25132f/CRM/Financial/BAO/PaymentProcessor.php#L328-328

$this->_processors is derived from 

    $processors = CRM_Financial_BAO_PaymentProcessor::getPaymentProcessors($capabilities);

& capabilities is set to array('LiveMode', 'BackOffice')

so, CRM_Financial_BAO_PaymentProcessor::getPaymentProcessors should not be returning test processors